geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/KafkaFeatureCacheWithMetrics.scala

```diff
@@ -0,0 +1,40 @@
+/***********************************************************************
+ * Copyright (c) 2013-2021 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php. 
+ ***********************************************************************/
+
+package org.locationtech.geomesa.kafka.index
+
+import com.codahale.metrics.Gauge
+import org.locationtech.geomesa.kafka.data.KafkaDataStore.IndexConfig
+import org.locationtech.geomesa.kafka.index.FeatureStateFactory.FeatureState
+import org.locationtech.geomesa.metrics.core.GeoMesaMetrics
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+class KafkaFeatureCacheWithMetrics(sft: SimpleFeatureType, config: IndexConfig, metrics: GeoMesaMetrics)
+    extends KafkaFeatureCacheImpl(sft, config){
+
+  metrics.register(sft.getTypeName, "index-size", new Gauge[Int] { override def getValue: Int = size() })
+
+  private val updates     = metrics.meter(sft.getTypeName, "updates")
+  private val removals    = metrics.meter(sft.getTypeName, "removals")
+  private val expirations = metrics.meter(sft.getTypeName, "expirations")
+
+  override def put(feature: SimpleFeature): Unit = {
```

Comment:
Is the use of 'clears' in the KDS going to throw things off?

Reply:
this is just counting the rate of incoming features, so clears shouldn't affect it

Reply:
Ok, so we do not have a metric that reports the number of records in the KDS cache, right?

Reply:
we do, it's the 'index-size' one a few lines up

Reply:
although this just reports the main hashmap size, the indices could potentially go out of sync. but i haven't seen that happen since the switch to using futures for expiring features.

Reply:
Got it.   As a sanity check, 'size()' calls `state.size()` (State is a ConcurrentHashMap.)

Here: https://github.com/locationtech/geomesa/blob/main/geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/KafkaFeatureCacheImpl.scala#L32 we have a comment about not iterating over `state`.

Is this the best way to compute the cache's size?

Reply:
size doesn't require iterating over the map. it's not just a lookup, but it should be essentially constant time: https://stackoverflow.com/a/22996395/7809538
it seems like they improved the algorithm in java 8 so that it doesn't ever need to lock, looking at the source for it.

Reply:
seems like there is some book-keeping done with each put/remove that simplifies the call to size